### PR TITLE
Allow py-pillow to link to Spack-built libraries

### PR DESCRIPTION
**Pull request by adamjstewart**
Previously, py-pillow linked to the system libraries instead of the Spack-built libraries. This PR fixes that.

Note that this solution is actually [recommended by the developers](https://pillow.readthedocs.io/en/latest/installation.html#external-libraries):
> If the prerequisites are installed in the standard library locations for your machine (e.g. `/usr` or `/usr/local`), no additional configuration should be required. If they are installed in a non-standard location, you may need to configure setuptools to use those locations by editing `setup.py` or `setup.cfg`, or by adding environment variables on the command line:
>
> ```$ CFLAGS="-I/usr/pkg/include" pip install pillow```

I actually tried using environment variables but to no avail. Hacking `setup.py` is also the [solution that Homebrew uses](https://github.com/Homebrew/homebrew-python/blob/master/pillow.rb).

@citibeth: You were the last person to edit py-pillow. Can you confirm that this solution works for you?

Note that I'm still having trouble getting it to link to Tcl/Tk. Any solutions are welcome.

### CODE REVIEW

**Comment by tgamblin** [comment]
I like it.  @lee218llnl: thoughts?

One nitpick: The variants `jpeg` and `jpeg2000` seem like they're not quite orthogonal.  If I build with `+jpeg+jpeg2000`, is `jpeg` used for regular JPEG and `openjpeg` used for JPEG 2000?  Or does it just use `openjpeg` in that case?

**Comment by lee218llnl** [comment]
@tgamblin looks good to me, although it looks like this package has evolved a lot from its original version. @citibeth should probably take a look.

**Comment by tgamblin** [comment]
@citibeth: are you ok with these changes?

**Comment by adamjstewart** [comment]
Not sure if it should be merged before Tcl/Tk support is fixed, but I guess it wasn't working before anyway. I also don't know the best way to resolve the last couple flake8 complaints.

**Comment by adamjstewart** [comment]
I actually figured out why it wasn't working before. We were previously specifying `--library-dirs (-L)` but not `--include-dirs (-I)`. Let me see if I can get Tcl/Tk working. The original method might be more flexible in that respect.

**Comment by adamjstewart** [comment]
> I like provides('PIL')

It's still there. I just moved it between the variants and the dependencies because I thought that was a more logical place to put it. I'll put it back. I know, it just makes reviewing this PR more confusing.

> Some systems (Ubuntu 12.04) have a very old version of binutils. depends_on('binutils') was added to packages, such as py-pillow, that won't build with the system binutils on these old systems. 

I'll add it back.

> The gratuitous whitespace changes are just annoying because they make it harder to review diffs. In this case, the PEP-8 style guide also argues against the result; so maybe better to put it back the way it was for both reasons.

I agree that this makes it harder to review diffs, but unaligned text messes with my OCD and makes the resulting code significantly less readable. It's the only PEP-8 recommendation that I disagree with. I'll defer the final decision to @tgamblin. I agree that we should be consistent on this one. I've seen a lot of packages that have bounced back and forth between aligned text and unaligned text with every PR.

> I'm not sure if '{}'.format() works with all the versions of Python we want Spack to work with (my attempts to use it have been rejected by Travis before on Python 2.6). So far, Spack has used old-style formatting throughout. These probably need to change back to old-style formatting, but @tgamblin would have the last word here.

From what I understand:

String Formatting Style | Python Versions Supported
--- | ---
`'%s' % x` | 2
`'{0}'.format(x)` | 2.6 - 2.7, 3
`'{}'.format(x)` | 2.7, 3

The old style only works for Python 2. Since the new style, `'{0}'.format(x)` works for all Python versions that Spack currently supports (2.6 and 2.7) and all future versions of Python, I believe it is the only string formatting syntax we should be using. Not that I'm going to call people out when they use the old syntax. But if I'm working on a PR, I'm going to update things to the new syntax while I'm at it. @tgamblin: Thoughts?

**Comment by citibeth** [comment]
Another issue... we need provides('pil'), not provides('PIL').  Package names are all-lowercase, and that includes virtual packages too.

**Comment by citibeth** [comment]
> It's still there. I just moved it between the variants and the dependencies because I thought that was a more logical place to put it. I'll put it back. I know, it just makes reviewing this PR more confusing.

Ooh... I thought it was newly added.  Either way, providing the virtual package seems like a good idea.

If it's not newly added, then that means we need to fix up any depends_on('PIL') to depends_on('pil').  If 'mpi' is good enough for Spack, then so should be 'pil'.

BTW... I noticed that libNBC seems to have slipped in with upper-case package name.

**Comment by citibeth** [comment]
> I agree that we should be consistent on this one. I've seen a lot of packages that have bounced back and forth between aligned text and unaligned text with every PR.

It's the bouncing that disturbs me the most.

**Comment by citibeth** [comment]
If the format() works with Python 2.6, then my objection to using it is
removed.  I knew that '{}'.format() didn't work, but did NOT know that
'{0}'.format() does work.  This issue is then in the same category as
whitespace changes.

**Comment by adamjstewart** [comment]
Submitted an issue regarding linking to Tcl/Tk: https://github.com/python-pillow/Pillow/issues/1972

**Comment by tgamblin** [comment]
@adamjstewart: so is this one ready to go?  it looks good to me.  Sounds like @citibeth is ok with it.

**Comment by adamjstewart** [comment]
@tgamblin: let's hold off for now. I want to see what the Pillow devs have to say first. I have a feeling that they'll think of some better way to link Tcl that uses `-L` and `-I` and I'll go back to the way @citibeth already had things. No reason for that much churn.

**Comment by tgamblin** [comment]
ok sounds good -- keep me posted.

**Comment by adamjstewart** [comment]
I spoke to the Pillow devs. It looks like they made a lot of new changes lately. Namely, the current master branch no longer accepts `--enable-tk` and `--enable-tcl`, and `setup.py` no longer contains `TCL_ROOT`. Pillow now uses whatever Tk/Tcl is available at run-time.

I think we should remove py-pillow's `+tk` variant. It doesn't do anything for the master branch and never worked for 3.0.0 and 3.2.0. If you want Tk/Tcl support for py-pillow, you'll just have to `spack install py-pillow@master ^python+tk`.

As far as the versions available, we have 3 options:

1. Add the master branch, drop 3.0.0 and 3.2.0.
2. Stay with 3.0.0 and 3.2.0 for now. Tk/Tcl support won't work until the next Pillow release comes out.
3. Support both and add `preferred=True` for the master branch.

Thoughts?

**Comment by citibeth** [comment]
> I think we should remove py-pillow's +tk variant. It doesn't do anything
> for the master branch and never worked for 3.0.0 and 3.2.0. If you want
> Tk/Tcl support for py-pillow, you'll just have to spack install
> py-pillow@master ^python+tk.
>

This is a good call.  You're not breaking any past Spack users, since +tk
wasn't available anyway.


> As far as the versions available, we have 3 options:
>
>    1. Add the master branch, drop 3.0.0 and 3.2.0.
>    2. Stay with 3.0.0 and 3.2.0 for now. Tk/Tcl support won't work until
>    the next Pillow release comes out.
>    3. Support both and add preferred=True for the master branch.
>
>
>
> Thoughts?
>

You should keep version 3.0.0 because that was there before you modified
the package.  And now that you've added 3.2.0, why not keep it as well...

It's also best to NOT set preferred=True on non-official releases, unless
there's a good reason.  The idea of "good reason" is fuzzy.  But you don't
want to break the build of others who rely on the package.  In this case,
I'd lean toward keeping 3.0.0 or 3.2.0 the preferred version: no bug or
security problem has been identified here, they just lack a feature you
want.

Finally... you can't just download from the master branch in a production
package.  Choose a specific version out of the package's git reo, and
specify that instead.  See below...

-- Elizabeth

Repo downloading is really convenient when you're testing / developing a
package.  For example, I have:
```
class PyGiss(Package):
    """Misc. Python Stuff."""

    homepage = "https://github.com/citibeth/pygiss"
    url      = "https://github.com/citibeth/pygiss/tarball/v0.1.0"

    version('0.1.2', '1c1c745c2818a6930c29c6ec7f835943')
    version('0.1.1', '172d468690a8b8f474884d7a60064bc7')
    version('glint2', git='https://github.com/citibeth/pygiss.git',
branch='glint2')
```

However, unchecksummed versions are a security risk for Spack users who are
NOT in control of the code in question.  Therefore, Spack users should
NEVER get them by default; they should have to work to get them.  I would
suggest the following in order of preference:
 1. If you use git versions for development, comment them out before
merging with Spack.
 2. If you don't comment them out, give them a non-numeric version number
and make sure they are NOT preferred=True.

In cases where we need to use a version of a package that's not been
officially released, we need to essentially do our own release of the
package for Spack users.  To do this, someone (the package.py author) needs
to find ONE version in the git repo that we like, and specify it, with
version number.  This is done by tickling GitHub with a URL that will get
it to produce a tarball of that version.  For example:

```
class PyProj(Package):
    """Python interface to the PROJ.4 Library."""
    homepage = "http://jswhit.github.io/pyproj/"
    url      = "https://github.com/jswhit/pyproj/tarball/v1.9.5.1rel"

    # This is not a tagged release of pyproj.
    # The changes in this "version" fix some bugs, especially with Python3
use.
    version('1.9.5.1.1', 'd035e4bc704d136db79b43ab371b27d2',
        url='
https://www.github.com/jswhit/pyproj/tarball/0be612cc9f972e38b50a90c946a9b353e2ab140f
')

    version('1.9.5.1', 'a4b80d7170fc82aee363d7f980279835')
```

Version "1.9.5.1.1" is not an official py-proj version, it's an interim
version.  I added an extra ".1" to the official version number 1.9.5.1 to
indicate this is something later than 1.9.5.1.  Two hash codes are
associated with this version: one is the git hash of the chosen version,
and one is the MD5 checksum for Spack.

If you're going to use a version of py-pillow that the authors haven't yet
released, I would recommend:
1. Follow the pattern above for py-proj.
2. Remove the interim version once an appropriate official version is
released.  (This is an exception to our normal Spack policy of not removing
past versions of things).
3. Don't make it preferred=True unless there's a good reason.

**Comment by adamjstewart** [comment]
Hmm, but what guarantee do I have that the current commit is any safer than whatever the latest commit is in the future? What if I put in a version for the master branch and keep it commented out? I could also just not put it in. It didn't support Tk/Tcl before, so it's no like anyone needs it right now.

**Comment by citibeth** [comment]
On Tue, Jun 21, 2016 at 1:30 PM, Adam J. Stewart <notifications@github.com>
wrote:

> Hmm, but what guarantee do I have that the current commit is any safer
> than whatever the latest commit is in the future?
>
If you use the checksummed tarball approach, we have the guarantee that
whatever version I see is the same as the one you saw when you made the
package.  We protect ourselves against malicious code being added to
py-pillow in the future, possibly unbeknownst to its authors.  That's all
we ever get with checksums, but it's a lot better than not using them.
Spack must conform to the security requirements of supercomputing centers,
and installing unchecksummed software is a huge violation of that
requirement.

Another problem with using a git branch is you don't even know if it
builds; or if it builds, whether it's been tested.  When someone complains
that they're using the @master version and it bombs, how can we help them
if their @master version is different from other Spack users' @master
version?


> What if I put in a version for the master branch and keep it commented
> out? I could also just not put it in. It didn't support Tk/Tcl before, so
> it's no like anyone needs it right now.
>
I would put in a master branch git download for yourself and try it out for
a while.  If that seems to be working out OK, then put put that exact git
hash you're using as a tarball download into the package, and share it in a
Spack PR.  We will then know that the version we're getting has been tested
out.

In other words... if we can't use an official release, someone on the Spack
end has to go through the QA required for a release of any software.

-- Elizabeth

**Comment by adamjstewart** [comment]
Alright, I removed the `+tk` variant, undid some of the gratuitous whitespace changes, and converted everything from `PIL` to `pil`. Both `py-basemap` and `py-scikit-image` previously explicitly depended on py-pillow. Now, they depend on the `pil` virtual package, which still defaults to `py-pillow`. The documentation for both packages lists PIL as a dependency, so I don't feel bad about having them depend on `pil`.

**Comment by citibeth** [comment]
Fine with me
